includes/controllers/class.llms.controller.achievements.php

```diff
@@ -26,48 +36,67 @@ class LLMS_Controller_Achievements {
 	 * @return void
 	 */
 	public function __construct() {
+
+		parent::__construct();
 		add_action( 'init', array( $this, 'maybe_handle_reporting_actions' ) );
 	}
 
 	/**
-	 * Handle achievement form actions to download (for students and admins) and to delete (admins only)
+	 * Returns a translated text of the given type.
 	 *
-	 * @since 3.18.0
-	 * @since 3.35.0 Sanitize `$_POST` data.
+	 * @since [version]
 	 *
-	 * @return void
+	 * @param int   $text_type One of the LLMS_Abstract_Controller_User_Engagements::TEXT_ constants.
+	 * @param array $variables Optional variables that are used in sprintf().
+	 * @return string
 	 */
-	public function maybe_handle_reporting_actions() {
-
-		if ( ! llms_verify_nonce( '_llms_achievement_actions_nonce', 'llms-achievement-actions' ) ) {
-			return;
-		}
+	protected function get_text( $text_type, $variables = array() ) {
 
-		if ( isset( $_POST['llms_delete_achievement'] ) ) {
-			$this->delete( llms_filter_input( INPUT_POST, 'achievement_id', FILTER_SANITIZE_NUMBER_INT ) );
+		switch ( $text_type ) {
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS:
+				return sprintf(
+					/* translators: %1$d: awarded achievement ID */
+					__( 'Sorry, you are not allowed to edit the awarded achievement #%1$d.', 'lifterlms' ),
+					( $variables['engagement_id'] ?? 0 )
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE:
+				return sprintf(
+					/* translators: %1$d: awarded achievement ID */
+					__( 'Sorry, the awarded achievement #%1$d does not have a valid achievement template.', 'lifterlms' ),
+					( $variables['engagement_id'] ?? 0 )
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENTS_INSUFFICIENT_PERMISSIONS:
+				return __( 'Sorry, you are not allowed to edit awarded achievements.', 'lifterlms' );
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENTS_INVALID_NONCE:
+				return __( 'Sorry, you are not allowed to sync awarded achievements.', 'lifterlms' );
+			case self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID:
+				return __( 'Sorry, you need to provide a valid awarded achievement ID.', 'lifterlms' );
+			case self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID:
+				return __( 'Sorry, you need to provide a valid achievement template ID.', 'lifterlms' );
+			default:
+				return parent::get_text( $text_type );
```

Comment:
@eri-trabiccolo, yes. The nearly identical `get_text()` methods are needed in `LLMS_Controller_Achievements` and `LLMS_Controller_Certificates` because translations should be of sentences and not of phrases. 

See https://github.com/gocodebox/lifterlms/pull/1946#issuecomment-1030073416.

I initially moved all the text to the abstract classes and injected the appropriate post type and singular or plural. It turned out to be bad for translators.

Reply:
I don't think we're in the sentence/phrase case here. Are we?
Also:
1) I think that `awarded achievement(s)` still is the post type label (plural or singular), so if we change it that string should change.
2) `It turned out to be bad for translators.`. How bad? I don't think it's worse, with your solution translators have to basically double the translations. I do understand that `awarded achievement` can be feminine and `awarded certificate` can be masculine in some languages, and some languages have feminine and masculine articles (Italian has feminine and masculine articles - both definite and indefinite), in this case is up to the translator, to me, to find a neutral solution OR to do something like "def/indef %1$s" where %1$s is the post type singular label.
This approach is used in wp core as well, e.g. https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2518 Or do you think this is a different case?

> I initially moved all the text to the abstract classes and injected the appropriate post type and singular or plural.

I think that was the right way to address this, less code, easier to maintain and dynamic.

Reply:
A big factor in the decision to abandon the work done on building text strings dynamically involved case folding. Take the [`llms_my_certificate` post type](https://github.com/gocodebox/lifterlms/blob/5.8.0/includes/class.llms.post-types.php#L790) for example. The names have each word capitalized, so we need to change the case depending on where it will be used in the sentence. For English, the first word in a sentence is usually capitalized. This rule is not used in all languages though. Worse yet, I found that in Greek, changing the case can change the spelling of a word. ["Awarded" = "Απονεμήθηκε", but "awarded" = "απονεμηθεί"](https://translate.google.com/?sl=en&tl=el&text=Awarded%20Certificate%0Aawarded%20certificate%0AAwarded%20certificate%0A&op=translate). Confusingly, changing the case of "Απονεμήθηκε Πιστοποιητικό" [seems to work](https://translate.google.com/?sl=el&tl=en&text=%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CE%A0%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%B1%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C&op=translate)?

Translating a bunch of similar text strings, while more work for the translators, allows for better translations.

Reply:
I understand your position. Sure your solution allows for better translations, but to me it's about balance.
How better? At what cost?
I gave you an example on how WordPress does basically the same thing I'm proposing in the rest api (an in other parts). 

You say:
> The names have each word capitalized, so we need to change the case depending on where it will be used in the sentence. For English, the first word in a sentence is usually capitalized.

Ok so the problem is that in our case we actually should "lowercase" the post type labels? Because something like
`'Sorry, you need to provide a valid Awarded Achievement ID.'` would be ugly?
Even if I understand the principle, we are talking about admin error messages that would show up like never or nearly never :D

And since we have to lowercase the post type labels, you found that...
> Worse yet, I found that in Greek, changing the case can change the spelling of a word. ["Awarded" = "Απονεμήθηκε", but "awarded" = "απονεμηθεί"](https://translate.google.com/?sl=en&tl=el&text=Awarded%20Certificate%0Aawarded%20certificate%0AAwarded%20certificate%0A&op=translate).

The real issue I find here is that we cannot use `strtolower()` with non utf8 charsets as the greek, but we should use [`mb_strtolower()`](https://www.php.net/manual/en/function.mb-strtolower.php) and I understand this is problematic as the `mb` suite is not built-in - but in this case we can just lowercase the string if the function exists and do nothing if it doesn't (rare cases).

What you say about changing the case in Greek can change the spelling of a word, is weird to me: you might be right, I don't know Greek and maybe you do, but certainly the link you provided doesn't prove your theory.
If you click on `απονεμηθεί πιστοποιητικό` translation you'll see that google suggests a variant:
`απονεμήθηκε πιστοποιητικό` were the first word is the lowercase version of `Απονεμήθηκε`.

I think that GoogleTranslate just tries to guess the "nature" of the first word, whether it's a name, an adjective (adverb) or a verb and for Google (but not necessarily for greek people who understand the context) the first capitalized word alters the "nature" of it.

You're basically saying that we cannot use the post type or taxonomy labels in localizable sentences (at least if we want to alter them with lowercase/uppercase), and that the mb_convert_case/mb_strtolower cannot be applied at least to the greek language.

Maybe you're right, I don't agree but it's fine. 
I don't think we have to spend more time on this: our opinion differ but there's nothing wrong with what you did so  I guess we can move on. :D

Reply:
@eri-trabiccolo wrote,
> I gave you an example on how WordPress does basically the same thing I'm proposing in the rest api (an in other parts).

I agree that injecting capitalized words in a REST API error message is not a problem. I'd be surprised if anyone translates those.

> Ok so the problem is that in our case we actually should "lowercase" the post type labels? Because something like
> `'Sorry, you need to provide a valid Awarded Achievement ID.'` would be ugly?
> Even if I understand the principle, we are talking about admin error messages that would show up like never or nearly never :D

I see your point. Having capitalized words in the middle of a sentence, while incorrect, is not that bad. In addition to admin error messages, this issue also affects [text in meta boxes](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php#L68). 

> The real issue I find here is that we cannot use strtolower() with non utf8 charsets as the greek, but we should use [mb_strtolower()](https://www.php.net/manual/en/function.mb-strtolower.php) and I understand this is problematic as the mb suite is not built-in - but in this case we can just lowercase the string if the function exists and do nothing if it doesn't (rare cases).

I wrote a class with functions to handle multibyte strings if the mb functions are available. They _usually_ are. If not, the built in string functions methods are used. I even wrote unit tests. I need to create a branch and push it up to the repo. 

> What you say about changing the case in Greek can change the spelling of a word, is weird to me: you might be right, I don't know Greek and maybe you do, but certainly the link you provided doesn't prove your theory.

The translator tools made me question whether changing the case could cause translation errors. I don't know the answer to that question, which is why, with a heavy heart,  I chose to abandon all the work I did on case conversion and building sentences in the abstract classes with concrete classes that return a full sentence.

> You're basically saying that we cannot use the post type or taxonomy labels in localizable sentences (at least if we want to alter them with lowercase/uppercase)

Yes, and I'm not very happy about it either.